src/cargo/core/resolver/mod.rs

```diff
@@ -307,21 +350,21 @@ pub fn resolve(summaries: &[(Summary, Method)],
 /// iterate through next.
 fn activate(cx: &mut Context,
             registry: &mut Registry,
-            parent: Option<&Rc<Summary>>,
+            parent: Option<&Summary>,
             candidate: Candidate,
             method: &Method)
             -> CargoResult<Option<DepsFrame>> {
     if let Some(parent) = parent {
-        cx.resolve_graph.link(parent.package_id().clone(),
-                              candidate.summary.package_id().clone());
+        cx.resolve_graph.push(GraphNode::Link(parent.package_id().clone(),
+                                           candidate.summary.package_id().clone()));
     }
 
     let activated = cx.flag_activated(&candidate.summary, method);
 
     let candidate = match candidate.replace {
         Some(replace) => {
-            cx.resolve_replacements.insert(candidate.summary.package_id().clone(),
-                                           replace.package_id().clone());
+            cx.resolve_replacements.push((candidate.summary.package_id().clone(),
```

Comment:
New `resolve_replacements` has slightly different behavior regarding duplicates. When we `.iter().filter()` it, we'll now get all the duplicates, instead of the only one. Does this matter? If there are no duplicates possible, let's add a comment to `resolve_replacements: RcList<(PackageId, PackageId)>,` explaining that it is effectively a map, and first elements must be unique.